File: tests/python3/src/T013_SchemaIDIterator.py
#!/usr/bin/env python3

import libgmsec_python3 as lp
from Test import Test


class Test_SchemaIDIterator(Test):

    def test(self):
        self.test_iterator()
        self.test_reset()


    def test_iterator(self):
        lp.log_info("Test has_next() and next()")

        self.test_schema_id_iterator_aux(lp.GMSEC_MSG_SPEC_CURRENT, lp.Specification.SchemaLevel_LEVEL_0, 73)
        self.test_schema_id_iterator_aux(lp.GMSEC_MSG_SPEC_CURRENT, lp.Specification.SchemaLevel_LEVEL_1, 104)
        self.test_schema_id_iterator_aux(lp.GMSEC_MSG_SPEC_CURRENT, lp.Specification.SchemaLevel_LEVEL_2, 107)

        self.test_schema_id_iterator_aux(lp.GMSEC_MSG_SPEC_2019_00, lp.Specification.SchemaLevel_LEVEL_0, 73)
        self.test_schema_id_iterator_aux(lp.GMSEC_MSG_SPEC_2019_00, lp.Specification.SchemaLevel_LEVEL_1, 104)
        self.test_schema_id_iterator_aux(lp.GMSEC_MSG_SPEC_2019_00, lp.Specification.SchemaLevel_LEVEL_2, 107)


    def test_reset(self):
        lp.log_info("Test reset()")

        spec = lp.Specification()

        iter = spec.get_schema_id_iterator()

        while iter.has_next():
            iter.next()

        self.check("Expected no more schema IDs", iter.has_next() == False)

        iter.reset()

        self.check("Expected to have schema IDs", iter.has_next())


    def test_schema_id_iterator_aux(self, specVersion, schemaLevel, expectedSchemaIds):
        config = lp.Config()

        config.add_value("gmsec-specification-version", "" + str(specVersion))
        config.add_value("gmsec-schema-level", "" + str(schemaLevel))

        spec = lp.Specification(config)

        iter = spec.get_schema_id_iterator()

        numSchemaIds = 0

        while iter.has_next():
            id = iter.next()

            self.check("Schema ID is null", id != None)

            numSchemaIds += 1

        self.check("Unexpected number of schema IDs", expectedSchemaIds == numSchemaIds)


Test.run('Test SchemaIDIterator', Test_SchemaIDIterator())

File: tests/perl/src/T013_SchemaIDIterator.pl
#!/usr/bin/env perl

use strict;
use Test;
use libgmsec_perl;


sub Test_SchemaIDIterator
{
	my ($test) = @_;

	$test->setDescription("Test SchemaIDIterator");

	test_iterator($test);
	test_reset($test);
}


sub test_iterator
{
	libgmsec_perl::logInfo("Test hasNext() and next()");

	my ($test) = @_;

	test_schema_id_iterator_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_CURRENT, $libgmsec_perl::Specification::SchemaLevel_LEVEL_0, 73);
	test_schema_id_iterator_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_CURRENT, $libgmsec_perl::Specification::SchemaLevel_LEVEL_1, 103);
	test_schema_id_iterator_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_CURRENT, $libgmsec_perl::Specification::SchemaLevel_LEVEL_2, 106);

	test_schema_id_iterator_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_2019_00, $libgmsec_perl::Specification::SchemaLevel_LEVEL_0, 73);
	test_schema_id_iterator_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_2019_00, $libgmsec_perl::Specification::SchemaLevel_LEVEL_1, 103);
	test_schema_id_iterator_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_2019_00, $libgmsec_perl::Specification::SchemaLevel_LEVEL_2, 106);
}


sub test_reset
{
	libgmsec_perl::logInfo("Test reset()");

	my ($test) = @_;

	my $spec = libgmsec_perl::Specification->new();

	my $iter = $spec->getSchemaIDIterator();

	while ($iter->hasNext())
	{
		$iter->next();
	}

	$test->check("Expected no more schema IDs", $iter->hasNext() == 0);

	$iter->reset();

	$test->check("Expected to have schema IDs", $iter->hasNext());
}


sub test_schema_id_iterator_aux
{
	my ($test, $specVersion, $schemaLevel, $expectedSchemaIds) = @_;

	my $config = libgmsec_perl::Config->new();
	$config->addValue("gmsec-specification-version", "$specVersion");
	$config->addValue("gmsec-schema-level", "$schemaLevel");

	my $spec = libgmsec_perl::Specification->new($config);

	my $iter = $spec->getSchemaIDIterator();

	my $numSchemaIds = 0;

	while ($iter->hasNext())
	{
		my $id = $iter->next();

		$test->check("Schema ID is null", $id != undef);

		$numSchemaIds += 1;
	}

	$test->check("Unexpected number of schema IDs", $expectedSchemaIds == $numSchemaIds);
}


Test::driver(\&Test_SchemaIDIterator);
